### DISCUSSION TITLE
Overload for put() OutputStream or allowing us to pass a Consumer<OutputStream>

### DISCUSSION THREAD

**Discussion by hakanai** [lookfirst/sardine#453] [Ideas]
Many APIs don't give back an `InputStream` for streaming serialisation, but instead ask me to pass an `OutputStream`. For instance, [GSON.toJson](https://google-gson.googlecode.com/svn/trunk/gson/docs/javadocs/com/google/gson/Gson.html#toJson%28java.lang.Object, java.lang.Appendable%29) does this, [GZIPOutputStream](http://docs.oracle.com/javase/7/docs/api/java/util/zip/GZIPOutputStream.html) does as well, but all the custom filesystem APIs we use at work do the same thing, because it's more flexible.

So I would like an overload to `put` where I can either do this:

```
    getSardine().put(childUri.toString(), stream -> {
        // I write to stream
    });
```

or this:

```
    try (OutputStream stream = getSardine().put(childUri.toString())) {
        // I write to stream
    }
```

My workaround for the lack of it will be to create a Pipe and spin up an additional thread to do the copy, which is a little awkward.

**Comment by lookfirst** [comment]
Seems sane. Pull request?

**Comment by hakanai** [comment]
OK I hacked together the bits of it that I do know how to do, but there is a missing piece. The method taking an InputStream reuses org.apache.http.entity.InputStreamEntity.

Obviously I have to make an alternative version of this which accepts WriteLogic, which would be vaguely like this:

```
public class WriteLogicEntity extends AbstractHttpEntity
{
    private final WriteLogic writeLogic;

    public InputStreamEntity(WriteLogic writeLogic)
    {
        this(instream, null);
    }

    public InputStreamEntity(WriteLogic writeLogic, ContentType contentType)
    {
        this.writeLogic = writeLogic;
        if (contentType != null)
        {
            this.setContentType(contentType.toString());
        }
    }

    public boolean isRepeatable()
    {
        // because we don't really know
        return false;
    }

    public long getContentLength()
    {
        // because we don't really know
        return -1L;
    }

    public InputStream getContent() throws IOException
    {
        //TODO: What do we do here?
    }

    public void writeTo(OutputStream outstream) throws IOException
    {
        writeLogic.write(outstream);
    }

    public boolean isStreaming()
    {
        return true;
    }
}
```

It seems one way or another, an HttpEntity has to have a method returning InputStream. It isn't clear from their Javadoc what I'm supposed to do in this situation.

isStreaming() is also questionable. They give an incomplete list of examples:
- Streamed entities that read data directly from the socket should return true.
- Self-contained entities should return false.
- Wrapping entities should delegate this call to the wrapped entity.

Of which my implementation is none of the above.

I think I'm going to submit a ticket to them to improve the Javadoc and/or the interface...

**Comment by lookfirst** [comment]
:+1:

**Comment by hakanai** [comment]
Because I forgot to mention the issue in the commit, pull request #194 adds this.

**Comment by hakanai** [comment]
Perhaps of interest, I was eventually forced to stop doing it this way with Sardine, reason being that some implementations of the WriteLogic interface took a long time before they sent any data, but Sardine opens the connection immediately because it doesn't know either - end result is the server sometimes kicked you off if it thought your connection had stalled.

So now we write to a temp file and send it up using `put(String,InputStream,Map)`. We can't use the method which takes `File` because (1) we use `Path` instead and (2) it doesn't have the `Map` overload which appears to be the only way to specify whether to overwrite or create a new file.
